python/ray/data/expressions.py

```diff
@@ -0,0 +1,181 @@
+from __future__ import annotations
+
+import operator
+from dataclasses import dataclass
+from typing import Any, Callable, Dict
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+import pyarrow.compute as pc
+
+from ray.util.annotations import DeveloperAPI
+
+
+# ──────────────────────────────────────
+#  Basic expression node definitions
+# ──────────────────────────────────────
+@DeveloperAPI
+class Expr:  # Base class – all expression nodes inherit from this
+    # Binary/boolean operator overloads
+    def _bin(self, other: Any, op: str) -> "Expr":
+        other = other if isinstance(other, Expr) else LiteralExpr(other)
+        return BinaryExpr(op, self, other)
+
+    # arithmetic
+    def __add__(self, other):
+        return self._bin(other, "add")
+
+    def __sub__(self, other):
+        return self._bin(other, "sub")
+
+    def __mul__(self, other):
+        return self._bin(other, "mul")
+
+    def __truediv__(self, other):
+        return self._bin(other, "div")
+
+    # comparison
+    def __gt__(self, other):
+        return self._bin(other, "gt")
+
+    def __lt__(self, other):
+        return self._bin(other, "lt")
+
+    def __ge__(self, other):
+        return self._bin(other, "ge")
+
+    def __le__(self, other):
+        return self._bin(other, "le")
+
+    def __eq__(self, other):
+        return self._bin(other, "eq")
+
+    # boolean
+    def __and__(self, other):
+        return self._bin(other, "and")
+
+    def __or__(self, other):
+        return self._bin(other, "or")
+
+    # Rename the output column
+    def alias(self, name: str) -> "AliasExpr":
+        return AliasExpr(self, name)
+
+
+@DeveloperAPI
+@dataclass(frozen=True, eq=False)
```

Comment:
We actually need an equality check to be able to compare the expressions eventually. 

We'd need to work this clash around somehow (internally we can use a method rather than operator)